File: benchmarks/linalg/python/matmul.py
#!/usr/bin/env python3
"""Numpy matmul benchmark. On macOS this uses vecLib (Accelerate);
elsewhere typically OpenBLAS or MKL."""

import argparse
import json
import time
import numpy as np

def main() -> None:
    p = argparse.ArgumentParser()
    p.add_argument("--N", type=int, default=256)
    p.add_argument("--K", type=int, default=100)
    args = p.parse_args()

    N, K = args.N, args.K

    rng = np.arange(N * N, dtype=np.float32)
    A = ((rng * 31 + 7) % 17).astype(np.float32) / 17.0
    B = ((rng * 13 + 3) % 19).astype(np.float32) / 19.0
    A = A.reshape(N, N)
    B = B.reshape(N, N)

    # Warm up.
    C = A @ B

    times: list[float] = []
    for _ in range(K):
        t0 = time.perf_counter()
        C = A @ B
        times.append((time.perf_counter() - t0) * 1000.0)
    times.sort()
    median_ms = times[K // 2]
    gflops = (2.0 * N * N * N) / (median_ms * 1e6)

    # Anti-DCE.
    _ = float(C[0, 0])

    print(json.dumps({
        "impl": "python-numpy",
        "N": N,
        "K": K,
        "median_ms": round(median_ms, 4),
        "gflops": round(gflops, 2),
        "blas": np.show_config_used() if hasattr(np, "show_config_used") else "unknown",
    }))

if __name__ == "__main__":
    main()

File: benchmarks/linalg/julia/matmul.jl
#!/usr/bin/env julia
# Julia matmul benchmark. Uses Julia's built-in `*` for matrices,
# which delegates to OpenBLAS by default. Run:
#   julia matmul.jl 256 100

using LinearAlgebra
using Statistics
using JSON

function main()
    N = length(ARGS) >= 1 ? parse(Int, ARGS[1]) : 256
    K = length(ARGS) >= 2 ? parse(Int, ARGS[2]) : 100

    idx = collect(Float32, 0:(N*N-1))
    A = reshape(Float32.((idx .* 31 .+ 7) .% 17 ./ 17), N, N)
    B = reshape(Float32.((idx .* 13 .+ 3) .% 19 ./ 19), N, N)

    # Warm up.
    C = A * B

    times = Float64[]
    for _ in 1:K
        t0 = time_ns()
        C = A * B
        push!(times, (time_ns() - t0) / 1e6)
    end
    sort!(times)
    median_ms = times[K ÷ 2 + 1]
    gflops = (2.0 * N * N * N) / (median_ms * 1e6)

    # Anti-DCE.
    _ = C[1, 1]

    println(JSON.json(Dict(
        "impl" => "julia",
        "N" => N,
        "K" => K,
        "median_ms" => round(median_ms; digits=4),
        "gflops" => round(gflops; digits=2),
        "blas" => string(BLAS.vendor()),
    )))
end

main()
